src/service/layers-impl.js

```diff
@@ -0,0 +1,225 @@
+ * size and position relative to its parent container.
+ */
+export class Layers {
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private @const {!./ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @const {!Window} */
+    this.win = ampdoc.win;
+
+    /** @private @const {!./viewport-impl.Viewport} */
+    this.viewport_ = viewportForDoc(ampdoc);
+  }
+
+  /**
+   * @param {!AmpElement} element
+   */
+  add(element) {
+    new LayerElement(element, this);
+  }
+
+  /**
+   * @param {!AmpElement} unusedElement
+   */
+  remove(unusedElement) {
+  }
+
+  getViewport() {
+    return this.viewport_;
+  }
+
+
+  /****************************
+   * Ferrying to LayerElement *
+   ****************************/
+
+  /**
+   * @param {!AmpElement} element
+   * @return {!../layout-rect.LayoutRectDef}
+   */
+  remeasure(element) {
+    return LayerElement.for(element).remeasure();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {!../layout-rect.LayoutRectDef}
+   */
+  getPageLayoutBox(element) {
+    return LayerElement.for(element).getPageLayoutBox();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {!../layout-rect.LayoutRectDef}
+   */
+  getAbsoluteLayoutBox(element) {
+    return LayerElement.for(element).getAbsoluteLayoutBox();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {boolean}
+   */
+  isFixed(element) {
+    return LayerElement.for(element).isFixed();
+  }
+}
+
+
+/**
+ * LayerElement (née Layout) caches and measures an element's layout rectangle.
+ */
+export class LayerElement {
+  /**
+   * @param {!AmpElement} element
+   * @return {!LayerElement}
+   */
+  static for(element) {
+    return /** @type {!LayerElement} */ (
+        dev().assert(LayerElement.forOptional(element),
+        'Missing layer prop on %s', element));
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {LayerElement}
+   */
+  static forOptional(element) {
+    return /** @type {LayerElement} */ (element[LAYER_PROP_]);
+  }
+
+
+  /**
+   * @param {!AmpElement} element
+   */
+  constructor(element, layers) {
+    element[LAYER_PROP_] = this;
+
+    /** @const {!AmpElement} */
+    this.element_ = element;
+
+    /** @const {!Layers} */
+    this.layers_ = layers;
+
+    /** @private {!../layout-rect.LayoutRectDef} */
+    this.layoutBox_ = layoutRectLtwh(-10000, -10000, 0, 0);
+
+    /** @private {boolean} */
+    this.isFixed_ = false;
+  }
+
+  /**
+   * Returns a previously measured layout box adjusted to absolute positioning.
+   * Note that fixed-position elements change absolute positioning based on the
+   * scroll position (they move with the scroll).
+   * @return {!../layout-rect.LayoutRectDef}
+   */
+  getAbsoluteLayoutBox() {
+    if (!this.isFixed()) {
+      return this.layoutBox_;
+    }
+
+    const viewport = this.layers_.getViewport();
+    return moveLayoutRect(this.layoutBox_, viewport.getScrollLeft(),
```

Comment:
Let's discuss. I'd like to get rid of BCR here and only use `offsetTop/Left/etc` for all calculations. And by extension get rid of most of concepts of Viewport around here.

I think I understand what you are trying to do here for transition. But I wonder if we may not be better getting by designing layers just the way we want and finding a way to either work Resources class by experiment or such?